### ISSUE TITLE
Decoder parameters become huge when scaling up the problem

### ISSUE THREAD

**Issue by danifranco** [ariG23498/mae-scalable-vision-learners#17]
Hello!

First thing I wanna thank you for this piece of code! Amazing work!

I have a question regarding the decoder implementation. In the last part the layers are defined as follows:

``` 
x = layers.Flatten()(x)
pre_final = layers.Dense(units=image_size * image_size * 3, activation="sigmoid")(x)
outputs = layers.Reshape((image_size, image_size, 3))(pre_final)
```

This does not correspond to the official implementation where they create a Linear layer as follows:

```
self.decoder_pred = nn.Linear(decoder_embed_dim, patch_size**2 * in_chans, bias=True) # decoder to patch
```

[Here is the link to the exac line](https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/models_mae.py#L58).

The problem is that when you try to scale your implementation up, to apply, for instance, images of 224x224 as in [ImageNet](https://www.image-net.org/) and configuring the same parameters as they used ([here one of the models they used, the Base one](https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/models_mae.py#L223)), this Dense layer becomes huge in parameters. It is even bigger than the encoder... You can see a model.summary() here that I made based in your code:
```
____________________________________________________________________
 Layer (type)                     Output Shape             Param #   
====================================================================
 patches (Patches)                 multiple                0                      
                                                                                                                                                      
 patch_encoder (PatchEncoder)      multiple                246784                 
                                                                                                                                                      
 mae_encoder (Functional)         (None, None, 768)        85056000               
                                                                                                                                                      
 mae_decoder (Functional)         (None, 128, 128, 1)      554104320              
                                                                                                                                                      
=====================================================================
Total params: 639,407,104
Trainable params: 639,407,104
```

As a tip also I tried the [notebook](https://colab.research.google.com/github/facebookresearch/mae/blob/main/demo/mae_visualize.ipynb) their provide in that official implementation and if print the model parameters the decoder has not the 10% of the parameters of the encoder, as they state in the manuscript. However, has a more reasonable parameter balance between encoder/decoder. See here an output of their model so you can check what should be:

```
from torchsummary import summary
summary(model_mae, (3, 224, 224))

OUTPUT:
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
================================================================
            Conv2d-1          [-1, 768, 14, 14]         590,592
        PatchEmbed-2             [-1, 196, 768]               0
         LayerNorm-3              [-1, 50, 768]           1,536
            Linear-4             [-1, 50, 2304]       1,771,776
           Dropout-5           [-1, 12, 50, 50]               0
            Linear-6              [-1, 50, 768]         590,592
           Dropout-7              [-1, 50, 768]               0
         Attention-8              [-1, 50, 768]               0
          Identity-9              [-1, 50, 768]               0
. . .
           Block-324             [-1, 197, 512]               0
       LayerNorm-325             [-1, 197, 512]           1,024
          Linear-326             [-1, 197, 768]         393,984
================================================================
Total params: 111,654,400
Trainable params: 111,654,400
Non-trainable params: 0
----------------------------------------------------------------
Input size (MB): 0.57
Forward/backward pass size (MB): 296.82
Params size (MB): 425.93
Estimated Total Size (MB): 723.33
----------------------------------------------------------------
```

**Comment by ariG23498** [comment]
Hey @danifranco

Thank you for taking interest in this work!

What I remember from our tutorial is that we tweaked the implementation in such a way that the decoder spits out the entire image all at once (hence the `image_size, image_size, 3`) that you have pointed out. This makes it easier for us to resuse the patchification layer and gather the patches from the output image.

```python
def calculate_loss(self, images, test=False):
    # Augment the input images.
    if test:
        augmeneted_images = self.test_augmentation_model(images)
    else:
        augmeneted_images = self.train_augmentation_model(images)

    # Patch the augmented images.
    patches = self.patch_layer(augmeneted_images)

    # Encode the patches.
    (
        unmasked_embeddings,
        masked_embeddings,
        unmasked_positions,
        mask_indices,
        unmask_indices,
    ) = self.patch_encoder(patches)

    # Pass the unmaksed patche to the encoder.
    encoder_outputs = self.encoder(unmasked_embeddings)

    # Create the decoder inputs.
    encoder_outputs = encoder_outputs + unmasked_positions
    decoder_inputs = tf.concat([encoder_outputs, masked_embeddings], axis=1)

    # Decode the inputs.
    decoder_outputs = self.decoder(decoder_inputs) # <----- outputs an image
    decoder_patches = self.patch_layer(decoder_outputs) # <----- patchifies the image and 

    loss_patch = tf.gather(patches, mask_indices, axis=1, batch_dims=1) # <----- gathers the patches for loss (this can be optimized a lot!)
    loss_output = tf.gather(decoder_patches, mask_indices, axis=1, batch_dims=1)

    # Compute the total loss.
    total_loss = self.compiled_loss(loss_patch, loss_output)

    return total_loss, loss_patch, loss_output
```

You are right here. We cannot scale this model up and it indeed is taking a hit on the parameter count. Please feel free to raise a PR if you wanted to solve this issue with better gathering and slicing techniques inside the training loop.

Please feel free to reach out if there are any more concerns.

**Comment by wolo-wolo** [comment]
Thank you for your great work. I have encountered the same problems, is the following code correct？

`    # -----------------ori-----
    # x = layers.Flatten()(x)
    # pre_final = layers.Dense(units=image_size * image_size * 3, activation="sigmoid")(x)
    # outputs = layers.Reshape((image_size, image_size, 3))(pre_final)
    # -----------------modified-----
    pre_final = layers.Dense(units=patch_size * patch_size * 3, activation="sigmoid")(x)
    outputs = layers.Reshape((image_size, image_size, 3))(pre_final)`

**Comment by wolo-wolo** [comment]
Thank you @ariG23498  for your great work. I have encountered the same problems, is the following code correct？
```
    # -----------------ori-----
    # x = layers.Flatten()(x)
    # pre_final = layers.Dense(units=image_size * image_size * 3, activation="sigmoid")(x)
    # outputs = layers.Reshape((image_size, image_size, 3))(pre_final)
    # -----------------modified-----
    pre_final = layers.Dense(units=patch_size * patch_size * 3, activation="sigmoid")(x)
    outputs = layers.Reshape((image_size, image_size, 3))(pre_final)
```

**Comment by ariG23498** [comment]
> Thank you @ariG23498 for your great work. I have encountered the same problems, is the following code correct？
> 
> ```
>     # -----------------ori-----
>     # x = layers.Flatten()(x)
>     # pre_final = layers.Dense(units=image_size * image_size * 3, activation="sigmoid")(x)
>     # outputs = layers.Reshape((image_size, image_size, 3))(pre_final)
>     # -----------------modified-----
>     pre_final = layers.Dense(units=patch_size * patch_size * 3, activation="sigmoid")(x)
>     outputs = layers.Reshape((image_size, image_size, 3))(pre_final)
> ```

It is correct, but as far as I remember (I might be wrong here as I had trained it quite some time ago), removing the `Flatten()` gave me poor results. The problem was with the dataset size and the number of epocs I guess.

You could probably give this a try and report the results in this issue. That would be a great contribution!

**Comment by ariG23498** [comment]
Closing this for inactivity. Please feel free to open it if the query is not resolved.
